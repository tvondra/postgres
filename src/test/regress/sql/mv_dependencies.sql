-- data type passed by value
CREATE TABLE functional_dependencies (
    a INT,
    b INT,
    c INT
);

-- unknown column
CREATE STATISTICS s1 WITH (dependencies) ON (unknown_column) FROM functional_dependencies;

-- single column
CREATE STATISTICS s1 WITH (dependencies) ON (a) FROM functional_dependencies;

-- single column, duplicated
CREATE STATISTICS s1 WITH (dependencies) ON (a,a) FROM functional_dependencies;

-- two columns, one duplicated
CREATE STATISTICS s1 WITH (dependencies) ON (a, a, b) FROM functional_dependencies;

-- correct command
CREATE STATISTICS s1 WITH (dependencies) ON (a, b, c) FROM functional_dependencies;

-- random data (no functional dependencies)
INSERT INTO functional_dependencies
     SELECT mod(i, 111), mod(i, 123), mod(i, 23) FROM generate_series(1,10000) s(i);

ANALYZE functional_dependencies;

SELECT staenabled, stadependencies
  FROM pg_statistic_ext WHERE starelid = 'functional_dependencies'::regclass;

TRUNCATE functional_dependencies;

-- a => b, a => c, b => c
INSERT INTO functional_dependencies
     SELECT i/10, i/100, i/200 FROM generate_series(1,10000) s(i);

ANALYZE functional_dependencies;

SELECT staenabled, stadependencies
  FROM pg_statistic_ext WHERE starelid = 'functional_dependencies'::regclass;

TRUNCATE functional_dependencies;

-- a => b, a => c
INSERT INTO functional_dependencies
     SELECT i/10, i/150, i/200 FROM generate_series(1,10000) s(i);
ANALYZE functional_dependencies;

SELECT staenabled, stadependencies
  FROM pg_statistic_ext WHERE starelid = 'functional_dependencies'::regclass;

TRUNCATE functional_dependencies;

-- a => b, a => c, b => c
-- check explain (expect bitmap index scan, not plain index scan)
INSERT INTO functional_dependencies
     SELECT mod(i,400), mod(i,200), mod(i,100) FROM generate_series(1,30000) s(i);

CREATE INDEX fdeps_idx ON functional_dependencies (a, b);

ANALYZE functional_dependencies;

SELECT staenabled, stadependencies
  FROM pg_statistic_ext WHERE starelid = 'functional_dependencies'::regclass;

EXPLAIN (COSTS off)
 SELECT * FROM functional_dependencies WHERE a = 10 AND b = 5;

DROP TABLE functional_dependencies;

-- varlena type (text)
CREATE TABLE functional_dependencies (
    a TEXT,
    b TEXT,
    c TEXT
);

CREATE STATISTICS s2 WITH (dependencies) ON (a, b, c) FROM functional_dependencies;

-- random data (no functional dependencies)
INSERT INTO functional_dependencies
     SELECT mod(i, 111), mod(i, 123), mod(i, 23) FROM generate_series(1,10000) s(i);

ANALYZE functional_dependencies;

SELECT staenabled, stadependencies
  FROM pg_statistic_ext WHERE starelid = 'functional_dependencies'::regclass;

TRUNCATE functional_dependencies;

-- a => b, a => c, b => c
INSERT INTO functional_dependencies
     SELECT i/10, i/100, i/200 FROM generate_series(1,10000) s(i);

ANALYZE functional_dependencies;

SELECT staenabled, stadependencies
  FROM pg_statistic_ext WHERE starelid = 'functional_dependencies'::regclass;

TRUNCATE functional_dependencies;

-- a => b, a => c
INSERT INTO functional_dependencies
     SELECT i/10, i/150, i/200 FROM generate_series(1,10000) s(i);

ANALYZE functional_dependencies;

SELECT staenabled, stadependencies
  FROM pg_statistic_ext WHERE starelid = 'functional_dependencies'::regclass;

TRUNCATE functional_dependencies;

-- a => b, a => c, b => c
-- check explain (expect bitmap index scan, not plain index scan)
INSERT INTO functional_dependencies
     SELECT mod(i,400), mod(i,200), mod(i,100) FROM generate_series(1,30000) s(i);

CREATE INDEX fdeps_idx ON functional_dependencies (a, b);

ANALYZE functional_dependencies;

SELECT staenabled, stadependencies
  FROM pg_statistic_ext WHERE starelid = 'functional_dependencies'::regclass;

EXPLAIN (COSTS off)
 SELECT * FROM functional_dependencies WHERE a = '10' AND b = '5';

DROP TABLE functional_dependencies;

-- NULL values (mix of int and text columns)
CREATE TABLE functional_dependencies (
    a INT,
    b TEXT,
    c INT,
    d TEXT
);

CREATE STATISTICS s3 WITH (dependencies) ON (a, b, c, d) FROM functional_dependencies;

INSERT INTO functional_dependencies
     SELECT
         mod(i, 100),
         (CASE WHEN mod(i, 200) = 0 THEN NULL ELSE mod(i,200) END),
         mod(i, 400),
         (CASE WHEN mod(i, 300) = 0 THEN NULL ELSE mod(i,600) END)
     FROM generate_series(1,10000) s(i);

ANALYZE functional_dependencies;

SELECT staenabled, stadependencies
  FROM pg_statistic_ext WHERE starelid = 'functional_dependencies'::regclass;

DROP TABLE functional_dependencies;
