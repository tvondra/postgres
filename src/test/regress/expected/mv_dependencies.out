-- data type passed by value
CREATE TABLE functional_dependencies (
    a INT,
    b INT,
    c INT
);
-- unknown column
CREATE STATISTICS s1 WITH (dependencies) ON (unknown_column) FROM functional_dependencies;
ERROR:  column "unknown_column" referenced in statistics does not exist
-- single column
CREATE STATISTICS s1 WITH (dependencies) ON (a) FROM functional_dependencies;
ERROR:  statistics require at least 2 columns
-- single column, duplicated
CREATE STATISTICS s1 WITH (dependencies) ON (a,a) FROM functional_dependencies;
ERROR:  duplicate column name in statistics definition
-- two columns, one duplicated
CREATE STATISTICS s1 WITH (dependencies) ON (a, a, b) FROM functional_dependencies;
ERROR:  duplicate column name in statistics definition
-- correct command
CREATE STATISTICS s1 WITH (dependencies) ON (a, b, c) FROM functional_dependencies;
-- random data (no functional dependencies)
INSERT INTO functional_dependencies
     SELECT mod(i, 111), mod(i, 123), mod(i, 23) FROM generate_series(1,10000) s(i);
ANALYZE functional_dependencies;
SELECT staenabled, stadependencies
  FROM pg_statistic_ext WHERE starelid = 'functional_dependencies'::regclass;
 staenabled | stadependencies 
------------+-----------------
 {f}        | 
(1 row)

TRUNCATE functional_dependencies;
-- a => b, a => c, b => c
INSERT INTO functional_dependencies
     SELECT i/10, i/100, i/200 FROM generate_series(1,10000) s(i);
ANALYZE functional_dependencies;
SELECT staenabled, stadependencies
  FROM pg_statistic_ext WHERE starelid = 'functional_dependencies'::regclass;
 staenabled |                                                 stadependencies                                                 
------------+-----------------------------------------------------------------------------------------------------------------
 {f}        | [{0 => 1 : 0.999900}, {0 => 2 : 0.999900}, {1 => 2 : 0.999900}, {0, 1 => 2 : 0.999900}, {0, 2 => 1 : 0.999900}]
(1 row)

TRUNCATE functional_dependencies;
-- a => b, a => c
INSERT INTO functional_dependencies
     SELECT i/10, i/150, i/200 FROM generate_series(1,10000) s(i);
ANALYZE functional_dependencies;
SELECT staenabled, stadependencies
  FROM pg_statistic_ext WHERE starelid = 'functional_dependencies'::regclass;
 staenabled |                                                 stadependencies                                                 
------------+-----------------------------------------------------------------------------------------------------------------
 {f}        | [{0 => 1 : 0.999900}, {0 => 2 : 0.999900}, {1 => 2 : 0.494900}, {0, 1 => 2 : 0.999900}, {0, 2 => 1 : 0.999900}]
(1 row)

TRUNCATE functional_dependencies;
-- a => b, a => c, b => c
-- check explain (expect bitmap index scan, not plain index scan)
INSERT INTO functional_dependencies
     SELECT mod(i,400), mod(i,200), mod(i,100) FROM generate_series(1,30000) s(i);
CREATE INDEX fdeps_idx ON functional_dependencies (a, b);
ANALYZE functional_dependencies;
SELECT staenabled, stadependencies
  FROM pg_statistic_ext WHERE starelid = 'functional_dependencies'::regclass;
 staenabled |                                                 stadependencies                                                 
------------+-----------------------------------------------------------------------------------------------------------------
 {f}        | [{0 => 1 : 1.000000}, {0 => 2 : 1.000000}, {1 => 2 : 1.000000}, {0, 1 => 2 : 1.000000}, {0, 2 => 1 : 1.000000}]
(1 row)

EXPLAIN (COSTS off)
 SELECT * FROM functional_dependencies WHERE a = 10 AND b = 5;
                 QUERY PLAN                  
---------------------------------------------
 Bitmap Heap Scan on functional_dependencies
   Recheck Cond: ((a = 10) AND (b = 5))
   ->  Bitmap Index Scan on fdeps_idx
         Index Cond: ((a = 10) AND (b = 5))
(4 rows)

DROP TABLE functional_dependencies;
-- varlena type (text)
CREATE TABLE functional_dependencies (
    a TEXT,
    b TEXT,
    c TEXT
);
CREATE STATISTICS s2 WITH (dependencies) ON (a, b, c) FROM functional_dependencies;
-- random data (no functional dependencies)
INSERT INTO functional_dependencies
     SELECT mod(i, 111), mod(i, 123), mod(i, 23) FROM generate_series(1,10000) s(i);
ANALYZE functional_dependencies;
SELECT staenabled, stadependencies
  FROM pg_statistic_ext WHERE starelid = 'functional_dependencies'::regclass;
 staenabled | stadependencies 
------------+-----------------
 {f}        | 
(1 row)

TRUNCATE functional_dependencies;
-- a => b, a => c, b => c
INSERT INTO functional_dependencies
     SELECT i/10, i/100, i/200 FROM generate_series(1,10000) s(i);
ANALYZE functional_dependencies;
SELECT staenabled, stadependencies
  FROM pg_statistic_ext WHERE starelid = 'functional_dependencies'::regclass;
 staenabled |                                                 stadependencies                                                 
------------+-----------------------------------------------------------------------------------------------------------------
 {f}        | [{0 => 1 : 0.999900}, {0 => 2 : 0.999900}, {1 => 2 : 0.999900}, {0, 1 => 2 : 0.999900}, {0, 2 => 1 : 0.999900}]
(1 row)

TRUNCATE functional_dependencies;
-- a => b, a => c
INSERT INTO functional_dependencies
     SELECT i/10, i/150, i/200 FROM generate_series(1,10000) s(i);
ANALYZE functional_dependencies;
SELECT staenabled, stadependencies
  FROM pg_statistic_ext WHERE starelid = 'functional_dependencies'::regclass;
 staenabled |                                                 stadependencies                                                 
------------+-----------------------------------------------------------------------------------------------------------------
 {f}        | [{0 => 1 : 0.999900}, {0 => 2 : 0.999900}, {1 => 2 : 0.494900}, {0, 1 => 2 : 0.999900}, {0, 2 => 1 : 0.999900}]
(1 row)

TRUNCATE functional_dependencies;
-- a => b, a => c, b => c
-- check explain (expect bitmap index scan, not plain index scan)
INSERT INTO functional_dependencies
     SELECT mod(i,400), mod(i,200), mod(i,100) FROM generate_series(1,30000) s(i);
CREATE INDEX fdeps_idx ON functional_dependencies (a, b);
ANALYZE functional_dependencies;
SELECT staenabled, stadependencies
  FROM pg_statistic_ext WHERE starelid = 'functional_dependencies'::regclass;
 staenabled |                                                 stadependencies                                                 
------------+-----------------------------------------------------------------------------------------------------------------
 {f}        | [{0 => 1 : 1.000000}, {0 => 2 : 1.000000}, {1 => 2 : 1.000000}, {0, 1 => 2 : 1.000000}, {0, 2 => 1 : 1.000000}]
(1 row)

EXPLAIN (COSTS off)
 SELECT * FROM functional_dependencies WHERE a = '10' AND b = '5';
                         QUERY PLAN                         
------------------------------------------------------------
 Bitmap Heap Scan on functional_dependencies
   Recheck Cond: ((a = '10'::text) AND (b = '5'::text))
   ->  Bitmap Index Scan on fdeps_idx
         Index Cond: ((a = '10'::text) AND (b = '5'::text))
(4 rows)

DROP TABLE functional_dependencies;
-- NULL values (mix of int and text columns)
CREATE TABLE functional_dependencies (
    a INT,
    b TEXT,
    c INT,
    d TEXT
);
CREATE STATISTICS s3 WITH (dependencies) ON (a, b, c, d) FROM functional_dependencies;
INSERT INTO functional_dependencies
     SELECT
         mod(i, 100),
         (CASE WHEN mod(i, 200) = 0 THEN NULL ELSE mod(i,200) END),
         mod(i, 400),
         (CASE WHEN mod(i, 300) = 0 THEN NULL ELSE mod(i,600) END)
     FROM generate_series(1,10000) s(i);
ANALYZE functional_dependencies;
SELECT staenabled, stadependencies
  FROM pg_statistic_ext WHERE starelid = 'functional_dependencies'::regclass;
 staenabled |                                                                                                                                                 stadependencies                                                                                                                                                 
------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {f}        | [{1 => 0 : 1.000000}, {2 => 0 : 1.000000}, {2 => 1 : 1.000000}, {3 => 0 : 1.000000}, {3 => 1 : 0.996700}, {0, 2 => 1 : 1.000000}, {0, 3 => 1 : 0.996700}, {1, 2 => 0 : 1.000000}, {1, 3 => 0 : 1.000000}, {2, 3 => 0 : 1.000000}, {2, 3 => 1 : 1.000000}, {0, 2, 3 => 1 : 1.000000}, {1, 2, 3 => 0 : 1.000000}]
(1 row)

DROP TABLE functional_dependencies;
